src/hotspot/share/opto/superword.cpp

```diff
@@ -3106,14 +3106,37 @@ VStatus VLoopBody::construct() {
     } else if (!post_visited.test(bb_idx(n))) {
       // cross or back arc
       const int old_length = stack.length();
+
+      // If a Load depends on the same memory state as a Store, we must make sure that
+      // the Load is ordered before the Store.
+      //
+      //      mem
+      //       |
+      //    +--+--+
+      //    |     |
+      //    |    Load (n)
+      //    |
+      //   Store (mem_use)
+      //
+      if (n->is_Load()) {
+        Node* mem = n->in(MemNode::Memory);
+        for (DUIterator_Fast imax, i = mem->fast_outs(imax); i < imax; i++) {
+          Node* mem_use = mem->fast_out(i);
+          if (_vloop.in_bb(mem_use) && !visited.test(bb_idx(mem_use)) && mem_use->is_Store()) {
```

Comment:
`mem_use->is_Store()` check is cheap and should be first. It will also help to skip other checks for Load node.

Reply:
Sure, I can do that :)